core/src/shred_fetch_stage.rs

```diff
@@ -61,6 +62,8 @@ impl ShredFetchStage {
         };
         let mut stats = ShredFetchStats::default();
 
+        let my_slots_to_repair_for_wen_restart =
+            slots_to_repair_for_wen_restart.map(|slots| slots.read().unwrap().clone());
```

Comment:
Hmm, do you mean I could pass in my_slots_to_repair_for_wen_restart to should_discard_shred? I need to eventually read out of the read lock, and I don't think slots_to_repair changes that fast so I need to read it per shred.

Reply:
like can you hold the read lock for the entire for loop since it's written to very infrequently?

Reply:
It is written infrequently but not that infrequent, at the beginning of the restart it should change. Generally I would prefer making a copy over holding a lock for a long time, this vector should not be too big.

Reply:
I misunderstood the code, I thought `modify_packets` was called inside a loop, but it is in fact the entire loop.

This approach doesn't allow `slots_to_repair_for_wen_restart` to be updated right? since you're cloning it on initialization, further writes won't be reflected in `should_discard_shred`.

I think you should pass the Arc to `should_discard_shred` and grab the read lock when filtering. If I understand correctly, the thread that needs the write lock is happy to wait since you're bottlenecked on replay and the 100ms sleep in repair anyway. 

Another approach is to make `slots_to_repair_for_wen_restart` a crossbeam channel, to allow more granular communication between the 2 components. This would save you the 16k slot move here `*slots_to_repair_for_wen_restart.write().unwrap() = filtered_slots;`.
However this would make the logic more complex, as you now need to compute and send the diff instead.

Reply:
Ahh, in my prototype I read the slot list at the beginning of each loop, thanks for catching that. In most cases this should not be a 16k slot list, because we truncate it at local root and filter out all slots we already have. I think making a copy at the beginning of the loop is better than obtaining read lock over each shred. But if you are really worried about the cost of vec copy, I can pass in only the (min_slot, max_slot), since the main purpose here is preventing wen_restart validators from absorbing newly generated blocks, so that they won't vote on something generated after the restart happens.

Reply:
> because we truncate it at local root 
local root is from before the failure right? or is this being updated as you repair? 

ya I think passing `(min_slot, max_slot)` is probably the simplest here, would prefer to not grab the lock and copy 100kb for each packet batch received.

Reply:
also what's the attack here - forcing validators to repair future blocks as they're being generated? can we just set a cap using the slot # at the time of entering this phase of `wen_restart`?

Reply:
The main concern is not spending any CPU/memory on blocks not needed for restart, if we keep consuming new blocks the validator could OOM before they finish restart.

Reply:
I see, is there a minimum vote % required to repair a block? Or is the 42% only for deciding the optimistic slot?

Reply:
42% is the line for repairing a block. Only blocks above 42% will be considered candidates of the optimistic slot.

Reply:
ah got it. I think an easier solution is to move up the check in `WindowService::verify_repair`.

If you're in `wen_restart`, any packets that come in must be `packet.meta().repair()`, have a parseable `repair_response::nonce` and correspond to something in `OutstandingShredRepairs`.

Check out how `window_service` does it, it already does the validation by sharing a lock with `repair_service`. I assume the intention here is to avoid the `sigverify` cost, so we could move that lock up to `shred_fetch_stage` for use in `wen_restart`

This way you don't have to pass around the `slots_to_repair_for_wen_restart`, and you have built in verification that all incoming shreds are responses to a repair request.

Reply:
Nice suggestion. I don't care too much about cost of shred_fetch and shred_verify, I care more about CPU spent on replay and OOM. So I moved the check to verify_repair.

Reply:
nice yeah that works.
although i'm unsure why the repair meta check is in window service to begin with. seems like we should be doing this before sigverify...

Reply:
If we have to do it, let's do that migration in a different PR though, it's beyond the scope of this PR.

Reply:
closing the loop here, i asked @jbiseda about the location of the check and he said:
> sigverify uses a thread pool to verify the packets, verify_repair is called by prune_shreds_invalid_repair holding a lock on outstanding_requests
flooding with old shreds should be mitigated by dedup
verify_repair will remove an entry from the outstanding request cache. we wouldn't want to remove an entry then have the packet fail sigverify